common/components/BalanceSidebar/TokenBalances/AddCustomTokenForm/DecimalField.tsx

```diff
@@ -0,0 +1,126 @@
+import React from 'react';
+import { Input } from 'components/ui';
+import Spinner from 'components/ui/Spinner';
+import ERC20 from 'libs/erc20';
+import { shepherdProvider } from 'libs/nodes';
+import { isPositiveIntegerOrZero } from 'libs/validators';
+import { translateRaw } from 'translations';
+import { Result } from 'nano-result';
+
+interface OwnProps {
+  address?: string;
+  onChange(decimals: Result<string>): void;
+}
+
+interface State {
+  userInput: string;
+  decimals: Result<string>;
+  autoDecimal: boolean;
+  addressToLoad?: string;
+  loading: boolean;
+}
+
+export class DecimalField extends React.Component<OwnProps, State> {
+  public static getDerivedStateFromProps(
```

Comment:
This repetitious logic (along with the checking for request on promise) might be a sign to use `renderProps` for composition -- thoughts?

Reply:
Agreed.